### PR TITLE
Add metrics plugin

### LINKED ISSUE #1650: Metrics plugin
Makes it possible to get metrics about the service.
The user can see service metrics with the help of the `serverless info` command.

**Metrics:**
- Function runtime
- Invocations
- Throttles
- Errors
- Percentile

**Questions:**
- Is this a global plugin or AWS related? (Do we need both?)
- Is this `service` wide of `function` wide?

What do you guys think @serverless/vip 

/cc @flomotlik @eahefnawy

### PR DESCRIPTION
**Pull request by pmuens**
## What did you implement:

Closes #1650

Add a metrics plugin so that specific metrics for a function can be retrieved.

## How did you implement it:

Add core `metrics` and AWS implementation of the `metrics` plugin.

## How can we verify it:

Deploy a service, invoke a function and then run `serverless metrics --function hello` to see the metrics of this function.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

File: docs/providers/aws/cli-reference/metrics.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+```bash
+serverless metrics --function hello
+```
+
+## Options
+
+- `--function` or `-f` The function you want to fetch the metrics for. **Required**
+- `--stage` or `-s` The stage you want to view the function metrics for. If not provided, the plugin will use the default stage listed in `serverless.yml`. If that doesn't exist either it'll just fetch the metrics from the `dev` stage.
+- `--region` or `-r` The region you want to view the function metrics for. If not provided, the plugin will use the default region listed in `serverless.yml`. If that doesn't exist either it'll just fetch the metrics from the `us-east-1` region.
+- `--startTime` A specific unit in time to start fetching metrics from (ie: `2010-10-20` or `1469705761`). Must be a valid ISO 8601 format.
```
Comment (mthenw) [inline docs/providers/aws/cli-reference/metrics.md]: `--startTime` should accept the same format as `serverless logs` so user should be able to do 
```
serverless metrics --startTime 24h
```

 to get stats from last 24h

Comment (mthenw) [inline docs/providers/aws/cli-reference/metrics.md]: And there is no info about default value

File: lib/plugins/aws/metrics/awsMetrics.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    // validate function exists in service
+    this.options.function = this.serverless.service.getFunction(this.options.function).name;
+
+    const today = new Date();
+    let yesterday = new Date();
+    yesterday = yesterday.setDate(yesterday.getDate() - 1);
+    yesterday = new Date(yesterday);
+
+    if (this.options.startTime) {
+      const since = (['m', 'h', 'd']
```
Comment (mthenw) [inline lib/plugins/aws/metrics/awsMetrics.js]: it's exactly the same logic as in logs. Can we extract it to some function and use it in those two places?

File: docs/providers/aws/cli-reference/metrics.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+```bash
+serverless metrics --function hello
+```
+
+## Options
+
+- `--function` or `-f` The function you want to fetch the metrics for. **Required**
+- `--stage` or `-s` The stage you want to view the function metrics for. If not provided, the plugin will use the default stage listed in `serverless.yml`. If that doesn't exist either it'll just fetch the metrics from the `dev` stage.
+- `--region` or `-r` The region you want to view the function metrics for. If not provided, the plugin will use the default region listed in `serverless.yml`. If that doesn't exist either it'll just fetch the metrics from the `us-east-1` region.
+- `--startTime` A specific unit in time to start fetching metrics from (ie: `2010-10-20`, `1469705761`, `30m` (30 minutes ago), `2h` (2 days ago) or `3d` (3 days ago)). Date formats should be written in ISO 8601. Defaults to yesterday (24h ago).
```
Comment (mthenw) [inline docs/providers/aws/cli-reference/metrics.md]: "Defaults to 24h ago."

File: lib/plugins/aws/metrics/awsMetrics.js (hunk 2 above)
Comment (pmuens) [inline lib/plugins/aws/metrics/awsMetrics.js]: Hmm. I modified it a little bit. The code which is used in the `metrics` plugin does a little less than the one in the `logs` plugin.

File: docs/README.md [hunk 4]
```diff
@@ -52,6 +52,7 @@ The Serverless Framework allows you to deploy auto-scaling, pay-per-execution, e
         <li><a href="./providers/aws/cli-reference/deploy.md">Deploy</a></li>
         <li><a href="./providers/aws/cli-reference/invoke.md">Invoke</a></li>
         <li><a href="./providers/aws/cli-reference/logs.md">Logs</a></li>
+        <li><a href="./providers/aws/cli-reference/metrics.md">Setup</a></li>
```
Comment (nikgraf) [inline docs/README.md]: Should be `Metrics`

File: lib/plugins/aws/metrics/awsMetrics.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        .then(this.getMetrics)
+        .then(this.showMetrics),
+    };
+  }
+
+  extendedValidate() {
+    this.validate();
+
+    // validate function exists in service
+    this.options.function = this.serverless.service.getFunction(this.options.function).name;
```
Comment (nikgraf) [inline lib/plugins/aws/metrics/awsMetrics.js]: The comment says `validate`. Is some kind of validation happening here?

Comment (pmuens) [inline lib/plugins/aws/metrics/awsMetrics.js]: Yes. This call to `getFunction` checks if the function is available in the service.

It's used throughout the codebase.

File: lib/plugins/metrics/metrics.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          stage: {
+            usage: 'Stage of the service',
+            shortcut: 's',
+          },
+          region: {
+            usage: 'Region of the service',
+            shortcut: 'r',
+          },
+          startTime: {
+            usage: 'Start time for the metrics retrieval',
```
Comment (nikgraf) [inline lib/plugins/metrics/metrics.js]: This should give some example on how to use like in the docs.

File: lib/plugins/metrics/metrics.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+          },
+          endTime: {
+            usage: 'End time for the metrics retrieval',
```
Comment (nikgraf) [inline lib/plugins/metrics/metrics.js]: This should give some example on how to use like in the docs.

File: lib/plugins/aws/metrics/awsMetrics.js (hunk 5 above)
Comment (nikgraf) [inline lib/plugins/aws/metrics/awsMetrics.js]: and what happens if that functions fails?

**Comment by nikgraf** [comment]
@pmuens I really liked your suggestion in the other issue https://github.com/serverless/serverless/issues/2810#issuecomment-264180750: Not sure if we need the serverless metrics and serverless metrics function commands. IMHO we can still use serverless metrics and if the user has not specified a function name then the plugin will return the metrics of all functions in the service. We can extend this easily later on.

Should we do it before merging this one?

**Comment by pmuens** [comment]
@nikgraf thanks for the in-depth review.

I'll update the PR tomorrow. We can also add the `serverless metrics` function in there.

File: lib/plugins/aws/metrics/awsMetrics.js (hunk 5 above)
Comment (pmuens) [inline lib/plugins/aws/metrics/awsMetrics.js]: https://github.com/serverless/serverless/blob/master/lib/classes/Service.js#L183

**Comment by pmuens** [comment]
Just pushed some updates for the PR.

I haven't updated the functionality that `serverless metrics` returns metrics for all function. Here's why:
- It still takes a significant amount of time as we still do 4 requests per function to get all the data
- I would assume that I see metrics of my whole service when I run `serverless metrics` but CloudWatch only supports metrics for Lambdas right now
- Displaying the metrics in a human readable way is a tricky thing as it'll end up in lots of scrolling if your service has lots of functions (which might be not unusual)
- The displaying problem gets even more problematic when we extend the metrics to display more info soon

I would suggest to open up a separate issue for this one so that we can write a proposal and discuss it thoroughly.

**Comment by nikgraf** [comment]
@pmuens sounds good to me

@mthenw @eahefnawy if you guys agree I would merge it

**Comment by mthenw** [comment]
LGTM

> I would suggest to open up a separate issue for this one so that we can write a proposal and discuss it thoroughly.

+1, I have some ideas

**Pull request merged**